nixos/lib/testing/matrix.nix

```diff
@@ -0,0 +1,132 @@
+let
+  inherit (lib) mkOption types;
+  inMatrixModule = choiceName: {
+    config = {
+      _matrixAttrsDone.${choiceName} = {};
+    };
+  };
+
+  choice = choice@{ name, ... }: {
+    options = {
+      after = mkOption {
+        description = ''
+          > You've already made the choice, now you have to understand it.
+
+          Some choices depend on other choices.
+
+          For example, we wouldn't be able to decide about the cookie before
+          we take the pill. Deciding the cookie is not possible without
+          crucial values set by the pill. To avoid an error like `The option
+          'pill' is used but not defined`, we force the ordering.
+
+          ```nix
+            matrix.cookie = {
+              after = ["pill"];
+            };
+          ```
+        '';
+        type = types.listOf types.str;
+        default = [];
+      };
+      enable = mkOption {
+        type = types.bool;
+        default = true;
+        description = ''
+          > But if you already know, how can I make a choice?
+
+          > Because you didn't come here to make the choice, you've already made it. You're here to try to understand why you made it.
+
+          Some choices don't make sense, because of earlier choices you've made.
+
+          With this option, you can make Nix understand that, pruning nonsensical
+          decisions from the returned tree of tests, so that we don't need to
+          compute more than necessary.
+
+          ```nix
+          matrix.cookie = {
+            after = ["pill"];
+            enable = config.pill == "red";
+            # ...
+          ```
+
+          If we chose the blue pill, we would never have to make a choice about
+          the cookie, so we skip this choice.
+        '';
+      };
+      value = mkOption {
+        type = types.lazyAttrsOf (extendModules { modules = [ (inMatrixModule choice.name) ]; }).type;
+      };
+    };
+  };
+in
+{
+  options = {
+    matrix = mkOption {
+      type = types.lazyAttrsOf (types.submodule choice);
+      description = ''
+        Wake up, Neo...
+
+        Like every other test, it is born into a prison that it cannot smell or taste or touch. A prison for the mind... Unfortunately, no one can be told what the matrix option is. You have to see it for yourself. This is your last chance. After this, there is no turning back.
+
+        You define
+
+        ```nix
+          matrix.pill.value.blue = { config, ... }: {
+            defaults.networking.domain = "bed.home.lan";
+          };
+        ```
+
+        and wake up in your bed and believe whatever you want to believe.
+
+        You define
+
+        ```nix
+          matrix.pill.value.red = { config, ... }: {
+            defaults.networking.domain = "wonderland.example.com";
+          };
+        ```
+
+        you test in wonderland, and I show you how deep the rabbit hole goes...
+
+        Remember, all I'm offering is the multiverse... Follow me...
```

Comment:
Just :heart:

Reply:
This is all fun, but should rather go to your personal blog post. Literally everyone outside the western cultural circle in their mid-30s will be completely thrown off. And possibly pissed to have to filter out what matters if they just want to get their job done.

(Please leave it in the Git history though, because it *is* awesome!)

Reply:
I agree. I already felt it was on the edge of acceptable, but you're absolutely right about the biases. Thank you for reminding me!
In a way it was also "hiding" the fact that that part of the documentation was rather incomplete. 

I haven't rebased it away, so you'll still be able to find it. (`Remove movie references`; hash might change)